### PR TITLE
NURBS patch 2D rotation functions [nurbs 2drot]

### PR DESCRIPTION
**Pull request by JacobLotz**
Split of PR https://github.com/mfem/mfem/pull/2724. Others will follow.

This PR adds a rotation function for 2D NURBS patches. This option was only available for 3D NURBS patches earlier.
It adds three functions to NURBSPatch in nurbs.cpp:
- `void Rotate2D(double angle); `
- `static void Get2DRotationMatrix(double angle, DenseMatrix &T);`
     --> These speak for them selves and are closely related to the 3D counterparts.
- `void Rotate(double angle, double normal[]= NULL);`  --> Streamlined function of the 2D and 3D rotation functions. The normal is not required for 2D rotations and is `NULL` by default. It has a check to make sure it is defined for 3D rotations.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3226](https://github.com/mfem/mfem/pull/3226) | @JacobLotz | @tzanio | @dylan-copeland + @v-dobrev | 10/26/22 | 4/29/23 | 5/6/23 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [x] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [x] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [x] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Comment by dylan-copeland** [comment]
@JacobLotz Could you please provide an example of the intended usage of the new features? For example, can this PR be used to rotate a NURBS mesh? Would that require a function in `NURBSExtension` to loop over all the patches?

**Comment by JacobLotz** [comment]
Hi @dylan-copeland, 

Usage of `Rotate` is in my opinion mainly in using MFEM to create a NURBS mesh. You will have to define the coordinates of your NURBS patches. For simulating flow around an airfoil for instance, you can use `Rotate` to rotate the coordinates around the origin and change the angle of attack of your airfoil in the mesh easily.

I was thinking about adding an example but I did not want to add too much code. Are you suggesting that I should add an example code to MFEM?

Currently you cannot use this PR to directly rotate a NURBS mesh. This would indeed require a loop over the patches in `NURBSExtension` and apply the rotation to each `NURBSPatch` in the extension.

**Comment by dylan-copeland** [comment]
Hi @JacobLotz, thank you for the explanation. It is acceptable to create a PR that only adds some utility functions without an example. If it is easy to make a small example, that would help the review process and to show users how to take advantage of the new features. Overall, MFEM does not have many NURBS examples, and of course this PR is not to blame. We should add more examples in `miniapps/nurbs`, especially for mesh generation. So far, MFEM only demonstrates how to use input NURBS meshes, without showing how to generate NURBS meshes. You are welcome to add some examples for your PRs, and I noticed one of your PRs does, but it is not required. We can proceed with the review, and you can simultaneously add examples if you wish.

File: mesh/nurbs.cpp [hunk 1]
```diff
@@ -1122,6 +1122,62 @@ void NURBSPatch::SwapDirections(int dir1, int dir2)
    swap(newpatch);
 }
 
+
```
Comment (dylan-copeland) [inline mesh/nurbs.cpp:1125]: Extra line

**Review by dylan-copeland — APPROVED** [review]

**Comment by JacobLotz** [comment]
Hi @dylan-copeland,
Thank you for you suggestions. I will think of other meshing examples I have available and I can add!

File: mesh/nurbs.cpp (hunk 1 above)
Comment (JacobLotz) [reply mesh/nurbs.cpp]: I have removed the extra line in [ede8ab9](https://github.com/mfem/mfem/pull/3226/commits/ede8ab9a0b61c7d3a8f54e5095b86a832525d87b)

**Comment by tzanio** [comment]
@IdoAkkerman and @v-dobrev, can you please review when you get a chance?

**Review by IdoAkkerman — APPROVED** [review]

**Comment by IdoAkkerman** [comment]
This augments the NURBS class with rotations for the 2D case. As such I approve.
However, it seems that the routine that computes the rotation matrix is not directly related with the NURBS class and could be placed outside (where??).
Furthermore, the rotation itself is also relatively standard, although the data access might be specific. So could the rotation itself also be placed in a generic mesh manipulation class?

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Review by tzanio — APPROVED** [review]
Thanks @JacobLotz !

**Pull request merged**